### PR TITLE
Feature - common library

### PR DESCRIPTION
**Pull request by yonikid15**
# Notes
- Moves all widely used resources into new common folder (components, helper functions, etc.)
- This library never needs to be compiled as a npm module, code from here is simply imported into other projects directly.
- The README in the common folder explains how to configure the aliases involved in importing.
- This PR touches nearly all the ts files in the repo - if a file wasn't moved to the common folder, some of its imports probably were.
- In the future, we may want to move even more code to the common folder. It makes sense to develop code in the most relevant project first, then move it to the common folder as needed.

### CODE REVIEW

**Comment by mnigh** [comment]
@yonikid15 what exactly is the "bug involving react-hook-form"?

**Comment by yonikid15** [comment]
> @yonikid15 what exactly is the "bug involving react-hook-form"?

https://github.com/GCTC-NTGC/gc-digital-talent/pull/431#issuecomment-912058045

**Review by tristan-orourke — COMMENTED** [review]
Cool! Nice job making this work.

What do you think about adding a bit more structure in how the common package exports things, based on the folder structure we had previously? So instead of 
```
import { navigate, Select } from "gc-digitial-talent-common";
```
it would be
```
import { navigate } from "gc-digitial-talent-common/helpers/router";
import { Select } from "gc-digital-talent-common/components/form";
```
I find the current flat structure of exports a little overwhelming and hard to parse. The exact structure could be open to discussion.

**Comment by tristan-orourke** [comment]
Also, should we consider moving Table (and GlobalFilter), Toast and Footer components to Common as well?

**Comment by yonikid15** [comment]
> Also, should we consider moving Table (and GlobalFilter), Toast and Footer components to Common as well?

Yes and possibly SideMenu

**Comment by tristan-orourke** [comment]
Well, I made a lot of changes so @yonikid15 and @mnigh you should both take a look again.

**Comment by tristan-orourke** [comment]
Also, awkwardly, this PR will almost certainly lead to merge conflicts, one way or another. Hopefully they won't be overwhelming.

File: admin/resources/js/components/GlobalFilter.tsx
Comment (mnigh) [inline admin/resources/js/components/GlobalFilter.tsx:27]: `preGlobalFilteredRows` is declared but its value is never read.

File: common/src/helpers/router.tsx
Comment (mnigh) [inline common/src/helpers/router.tsx:77]: `stripTrailingSlash` is declared but its value is never read. Is there a plan to use this function in the future?

File: admin/resources/js/components/PoolDashboard.tsx
Comment (mnigh) [inline admin/resources/js/components/PoolDashboard.tsx:88]: Remove `console.log`

File: common/src/components/Button/Button.stories.tsx
Comment (mnigh) [inline common/src/components/Button/Button.stories.tsx:10]: Is there another issue for creating all of the different button variations in storybook? These variations used to exist in storybook.

**Review by mnigh — CHANGES_REQUESTED** [review]
Some heavy lifting here, nice work @yonikid15 and @tristan-orourke 💪 ! I've posed several questions and pointed out some changes that are needed though before approval.

File: common/package.json
Comment (tristan-orourke) [reply common/package.json:6]: That's fine I think. `eslint-config-airbnb` is a real dependency, but eslint-*plugin*-airbnb is something I added accidentally at one point.

File: common/package.json
Comment (tristan-orourke) [reply common/package.json:14]: Yes that's on purpose. The h2- scripts in the common folder only affect the Storybook stories in /common folder. We still need to run h2- scripts in *admin* in order to build *admin* correctly.

Actually... We might need to add something to the h2 config file in /admin to ensure it searches through the common folder dependencies when building. @yonikid15 can you look into that?

File: common/src/helpers/router.tsx
Comment (tristan-orourke) [reply common/src/helpers/router.tsx:77]: Uh I guess not? We can probably remove it along with baseUrl()

File: common/src/components/Button/Button.stories.tsx
Comment (tristan-orourke) [reply common/src/components/Button/Button.stories.tsx:10]: I dunno what happened to them :/

File: common/package.json
Comment (mnigh) [reply common/package.json:6]: Cool. Worth running `npm run i` again and committing the updated package-lock.json file.

File: common/src/components/Button/Button.stories.tsx
Comment (yonikid15) [reply common/src/components/Button/Button.stories.tsx:10]: That's my fault, I forgot to replace it.

File: admin/hydrogen.config.json
Comment (tristan-orourke) [inline admin/hydrogen.config.json:6]: I would leave it at ""../common/src" just to to be safe.

File: common/hydrogen.config.json
Comment (tristan-orourke) [inline common/hydrogen.config.json:5]: Oops, this goes for the common folder too

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
admin\resources\js\components\Table.tsx references tries to pass in a preGlobalFilteredRows prop to GlobalFilter, and the prop doesn't exist anymore.

Besides that, I think this is ready to go.

**Review by tristan-orourke — APPROVED** [review]

File: admin/resources/js/components/Table.tsx
Comment (mnigh) [inline admin/resources/js/components/Table.tsx:85]: With this removal, [line 60 of this file](https://github.com/GCTC-NTGC/gc-digital-talent/pull/431/files/bb72ce5387b7d65bcb64665a0a71ac2485d09b31..eba6eba27731a7e46c6e64187c943d0c8d6bf565#diff-e1e060ddd846c8f9ab91f75034694e1bfe31685ba6471c1379b71b5841f8c757L60) now has: `preGlobalFilteredRows` is declared but its value is never read. Should be removed there as well.

**Review by mnigh — CHANGES_REQUESTED** [review]
Almost there! In addition to the one change suggested, `npm run i` should be run in the admin folder and the resulting changes to package-lock.json committed.

**Pull request merged**
